### DISCUSSION TITLE
How to use in paths unsupported tags from file?

### DISCUSSION THREAD

**Discussion by sylph1o** [beetbox/beets#4517] [Q&A] [answered]
Hi everyone!

I discovered recently this wonderful piece of FOSS and I am very excited to migrate my current library to beets. I have been looking at the feasibility of this, and particularly at the best way to reproduce my file organisation with beets.

I have discovered (from `beet fields`, the [documentation](https://beets.readthedocs.io/en/latest/reference/pathformat.html#available-values) and the [code](https://github.com/beetbox/mediafile/blob/master/mediafile.py#L1771-L2305)) that some metadata fields I use are not supported, although they are supported by MusicBrainz; e.g.
- [`albumsort`](https://picard-docs.musicbrainz.org/en/appendices/tag_mapping.html#album-sort-order-4),
- [`titlesort`](https://picard-docs.musicbrainz.org/en/appendices/tag_mapping.html#track-title-sort-order-4),
- [`originalartist`](https://picard-docs.musicbrainz.org/en/appendices/tag_mapping.html#original-artist).

I understand that `originalartist` is not very common. However, I am surprised that `albumsort` and `titlesort` are not supported, while `artistsort` and `albumartistsort` are. I currently rely on the values stored `albumsort` and `titlesort` (in the files) for the paths of my music library. I have found no mention of these fields, neither in the discussion board nor this Github project, so the feature must not be in high-demand. I am looking for a way to work around this limitation and be able to use the values stored in these fields in the paths defined by beets, preferably with minimal effort.

I thought flexible attributes could be a solution, using something like `beet import --set albumsort=$albumsort`, but unsupported tags stored in the files are simply not accessible from beets (not even during import). Is there a way for me to provide this value for each file from an external tools? For example, the following shell command return the value stored in the `albumsort` field of the file, but I do not know how to let beets read that _for each file_. 
```shell
ffprobe -v 0 -select_streams a:0 -show_entries stream_tags=albumsort /path/to/file.ogg | grep TAG | sed "s/TAG:albumsort=//g"
```

I guess I could write a shell script to loop over all files, read the tag with `ffprobe` as above and set a flexible attribute with `beet modify`, but that seems inefficient.
Does anyone have an idea to contribute? Any thought is welcome!

**Comment by sampsyo** [comment, marked as answer]
Good question! The way to add support for custom on-disk tags is to write a (very small) plugin that provides a media file specification:
https://beets.readthedocs.io/en/stable/dev/plugins.html#extend-mediafile

This way, you can provide beets field names for any metadata tags you like.

**Reply by sylph1o** [reply]
Thanks, this addresses my needs perfectly! I was able to add support for the tags I want, and to define the template fields I want to use in my paths. The plugin interface and good documentation made it surprisingly easy, so I'm grateful for that too.

Although my question has been answered, I have two remarks to add as feedback.

First, the plugin documentation explains how to use `pkgutil` to place the plugin (a Python module) inside the namespace package `beetsplug` and refers to [this question on StackOverflow](https://stackoverflow.com/q/1675734). The top-level answer there explains that `pkgutil` is no longer necessary since Python 3.3 introduced implicit namespace packages. I don't know if there is a reason why the documentation of beets still recommends to use `pkgutil`. I can only say that didn't use it and my plugin was loaded without issue.

Second, after loading my plugin, beets is now able to import the new metadata tags (e.g. `titlesort`) in its database when running `beet import <path>`. However, beets throws an error when I try to update an entry already existing in the database (i.e. with a `titlesort` tag in the file that was not read at import) with
```
beet update -F titlesort <query>
```
The error trace is copied below.
```
Traceback (most recent call last):
  File "/usr/sbin/beet", line 33, in <module>
    sys.exit(load_entry_point('beets==1.6.0', 'console_scripts', 'beet')())
  File "/usr/lib/python3.10/site-packages/beets/ui/__init__.py", line 1285, in main
    _raw_main(args)
  File "/usr/lib/python3.10/site-packages/beets/ui/__init__.py", line 1272, in _raw_main
    subcommand.func(lib, suboptions, subargs)
  File "/usr/lib/python3.10/site-packages/beets/ui/commands.py", line 1211, in update_func
    update_items(lib, decargs(args), opts.album, ui.should_move(opts.move),
  File "/usr/lib/python3.10/site-packages/beets/ui/commands.py", line 1163, in update_items
    item.store(fields=fields)
  File "/usr/lib/python3.10/site-packages/beets/library.py", line 344, in store
    super().store(fields)
  File "/usr/lib/python3.10/site-packages/beets/dbcore/db.py", line 528, in store
    tx.mutate(query, subvars)
  File "/usr/lib/python3.10/site-packages/beets/dbcore/db.py", line 866, in mutate
    cursor = self.db._connection().execute(statement, subvals)
sqlite3.OperationalError: no such column: titlesort
```
This is not really a problem for me: I can simply remove the entry and import it again. I'm simply reporting the unexpected behaviour. Would you like me to open an issue about it?

**Reply by sampsyo** [reply]
It's a great idea to remove the old documentation about `pkgutil`. We should almost certainly just delete this. If you have a chance, a PR doing just this would be extremely helpful.

The crash when doing `beet update` looks like it could be a bug! It looks like this code is expecting to support flexible attributes (it only supports fixed attributes, which correspond directly to SQLite columns). Maybe we should take this to a separate GitHub issue to investigate.
